examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+
+        if not isinstance(limit_val_batches, int):
+            assert limit_val_batches == float("inf")
+
+        self.limit_train_batches = limit_train_batches
+        self.limit_val_batches = limit_val_batches
+        self.val_sanity_check = val_sanity_check
+        self.validation_frequency = validation_frequency
+        self.use_distributed_sampler = use_distributed_sampler
+        self._current_train_return: Union[torch.Tensor, Mapping[str, Any]] = {}
+        self._current_val_return: Optional[Union[torch.Tensor, Mapping[str, Any]]] = {}
+
+        self.checkpoint_dir = checkpoint_dir
+        self.checkpoint_frequency = checkpoint_frequency
+
+    def fit(
+        self, model: LightningModule, train_loader: DataLoader, val_loader: DataLoader, ckpt_path: Optional[str] = None
+    ):
+        if self.fabric.is_global_zero:
+            self.fabric.call("prepare_data")
+
+        # TODO: Have fabric launch the loop function directly?
+        self.fabric.launch()
+
+        # setup dataloaders
+        train_loader = self.fabric.setup_dataloaders(train_loader, use_distributed_sampler=self.use_distributed_sampler)
+        if val_loader is not None:
+            val_loader = self.fabric.setup_dataloaders(val_loader, use_distributed_sampler=self.use_distributed_sampler)
+
+        # setup model and optimizer
+        if isinstance(self.fabric.strategy, FSDPStrategy):
+            model = self.fabric.setup_module(model)
+            # FIXME: currently, there is no way to support fsdp with model.configure_optimizers in fabric
+            # as it would require fabric to hold a reference to the model, which we don't want to.
+            raise NotImplementedError("BYOT currently does not support FSDP")
+            optimizer, scheduler_cfg = self._parse_optimizers_schedulers(model.configure_optimizers())
+            optimizer = self.fabric.setup_optimizers(optimizer)
+        else:
+            optimizer, scheduler_cfg = self._parse_optimizers_schedulers(model.configure_optimizers())
+            assert optimizer is not None
+            model, optimizer = self.fabric.setup(model, optimizer)
+
+        # assemble state (current epoch and global step will be added in save)
+        state = {"model": model, "optim": optimizer, "scheduler": scheduler_cfg}
+
+        # load last checkpoint if available
+        latest_checkpoint_path = self.get_latest_checkpoint(self.checkpoint_dir)
+        if latest_checkpoint_path is not None:
+            self.load(state, latest_checkpoint_path)
+
+            # check if we even need to train here
+            if self.max_epochs is not None and self.current_epoch >= self.max_epochs:
+                self.should_stop = True
+
+        # Run validation sanity check
+        if self.val_sanity_check and val_loader is not None and not self.should_stop:
+            self.val_loop(model, val_loader, limit_batches=self.val_sanity_check)
+
+        # TODO: should this be a for loop?
```

Comment:
@awaelchli @carmocca @lantiga opinions?

Reply:
If it makes things more readable, sure!

Reply:
it makes things more readable in terms of the loop itself, but makes following the stopping conditions way harder, IMO